core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -54,5 +77,228 @@ func (ft *GCSFileTransfer) Upload(task *ReferenceArtifactUploadTask) error {
 
-	return fmt.Errorf("not implemented yet")
+	reference := task.Reference
+
+	// Parse the reference path to get the scheme, bucket, and object
+	bucketName, objectName, err := parseReference(reference)
+	if err != nil {
+		return formatDownloadError("error parsing reference", err)
+	}
+	bucket := ft.client.Bucket(bucketName)
+
+	// If not using checksum, we set the digest to the reference
+	if task.Digest == reference {
+		// Get all of the objects under the reference
+		var objectNames []string
+		query := &storage.Query{Prefix: objectName}
+		it := bucket.Objects(ft.ctx, query)
+		for {
+			objAttrs, err := it.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return formatDownloadError("error iterating through objects", err)
+			}
+			objectNames = append(objectNames, objAttrs.Name)
+		}
+
+		// Try to download all objects under the reference
+		g := new(errgroup.Group)
+		g.SetLimit(maxWorkers)
+		for _, name := range objectNames {
+			objName := name // for closure in the goroutine
+			g.Go(func() error {
+				object, _, err := ft.getObjectAndAttrs(bucket, task, objName)
+				if err != nil {
+					if errors.Is(err, ErrObjectIsDirectory) {
+						ft.logger.Debug(
+							"GCSFileTransfer: Download: skipping reference because it seems to be a folder",
+							"reference", reference,
+							"object", objName,
+						)
+						return nil
+					}
+					return formatDownloadError(fmt.Sprintf("error getting object %s", objName), err)
+				}
+				localPath := getDownloadFilePath(objName, objectName, task.Path)
+				return ft.downloadFile(object, localPath)
+			})
+		}
+
+		// Wait for all of the go routines to complete and return an error if any errored out
+		if err := g.Wait(); err != nil {
+			return formatDownloadError("", err)
+		}
+	} else {
+		object, objAttrs, err := ft.getObjectAndAttrs(bucket, task, objectName)
+		if err != nil {
+			if errors.Is(err, ErrObjectIsDirectory) {
+				ft.logger.Debug(
+					"GCSFileTransfer: Download: skipping reference because it seems to be a folder",
+					"reference", reference,
+				)
+				return nil
+			}
+			return formatDownloadError(fmt.Sprintf("error getting object %s", objectName), err)
+		}
+		if objAttrs.Etag != task.Digest {
+			err := fmt.Errorf("digest/etag mismatch: etag %s does not match expected digest %s", objAttrs.Etag, task.Digest)
+			return formatDownloadError("", err)
+		}
+		return ft.downloadFile(object, task.Path)
+	}
+	return nil
+}
+
+// DownloadFile downloads the contents of object into a file at path localPath
+func (ft *GCSFileTransfer) downloadFile(object *storage.ObjectHandle, localPath string) error {
+	r, err := object.NewReader(ft.ctx)
+	if err != nil {
+		return formatDownloadError("error creating reader", err)
+	}
+	defer func(file io.ReadCloser) {
+		if err := file.Close(); err != nil {
+			ft.logger.CaptureError(
+				fmt.Errorf(
+					"file transfer: download: error closing response reader: %v",
+					err,
+				))
+		}
+	}(r)
```

Comment:
I believe you can ignore the Close() error here since you're only reading data. It would matter if you were *writing* data, since in that case a Close() could indicate that data wasn't successfully flushed; but since you're reading and all the Read() calls succeeded, there are no remaining failure modes.

```suggestion
	defer r.Close()
```